elyra/validation/validate.py

```diff
@@ -0,0 +1,424 @@
+        for isolate in nx.isolates(graph):
+            response.add_message(severity='2',
+                                 message_type="singletonReference",
+                                 message="This node is not connected to any part of the pipeline",
+                                 data={"nodeID": isolate,
+                                       "pipelineID": pipeline_id})
+
+        # get all cycles detected in the graph
+        for cycle in nx.simple_cycles(graph):
+            if len(cycle) == 1:
+                for node in node_list:
+                    for link in node['inputs'][0]['links']:
+                        if cycle[0] == link['node_id_ref']:
+                            response.add_message(severity='1',
+                                                 message_type="circularReference",
+                                                 message="A cycle was found within this pipeline",
+                                                 data={"nodeID": node['id'],
+                                                       "pipelineID": pipeline_id})
+            else:
+                for node in node_list:
+                    for link in node['inputs'][0]['links']:
+                        if cycle[0] == link['node_id_ref'] and cycle[1] == node['id']:
+                            response.add_message(severity='1',
+                                                 message_type="circularReference",
+                                                 message="A cycle was found within this pipeline",
+                                                 data={"nodeID": node['id'],
+                                                       "pipelineID": pipeline_id})
+
+
+class ValidationResponse(object):
+    def __init__(self):
+        self._response = {'title': "Elyra Pipeline Diagnostics",
+                          'description': "Issues discovered when parsing the pipeline",
+                          'issues': []
+                          }
+
+    @property
+    def response(self) -> Dict:
+        """
+        :return: The dict of validation errors and warnings found in the pipeline
+        """
+        return self._response
+
+    def add_message(self,
+                    message: str,
+                    message_type: Optional[str] = "",
+                    data: Optional[Dict[str, str]] = "",
+                    severity: str = '2'):
+        """
+        Helper function to add a diagnostic message to the response to be sent back
+        :param message: A simple message describing the issue
+        :param message_type: The type of message to send back e.g. invalidNodeType, invalidPipeline
+        :param data: granular details regarding the error e.g. the nodeID, pipelineID, linkID etc.
+        :param severity: the severity of the issue on a scale of 1-4
+        :return:
+        """
+        valid_severity_levels = ['1', '2', '3', '4']  # error | warning | information | hint
```

Comment:
I think it would be nice to use an enum for these values so that the enumerated value could be used rather than the stringized integer everywhere.
```python
from enum import Enum
class ValidationSeverity(Enum):
    Error = '1'
    Warning = '2'
    Information = '3'
    Hint = '4'
```
```suggestion
        valid_severity_levels = [ValidationSeverity.Error, ValidationSeverity.Warning, ValidationSeverity.Information, ValidationSeverity.Hint]
```
You would likely need to use `str(severity.value)` below if you truly wanted, say, "1" to be sent as the value.

Is there a reason these are string values and not ints?  If the later, you could derive from `IntEnum` and go with `severity.value`.

At any rate, using enum may not work out, but we should define "constants" for the numeric strings at a minimum.